### PR TITLE
More granular & configurable Cabal discovery

### PR DESCRIPTION
**Pull request by ramirez7**
https://github.com/dustin/graphex/issues/28

```
$ graphex cabal --help
Usage: graphex cabal [--discover-all | --discover-all-libs | 
                       --discover-all-exes | --discover-all-tests | 
                       --no-discover-all-libs | --no-discover-all-exes | 
                       --no-discover-all-exes | --discover-lib | 
                       --no-discover-lib | --discover-sublib ARG | 
                       --no-discover-sublib ARG | --discover-exe ARG | 
                       --no-discover-exe ARG | --discover-test ARG | 
                       --no-discover-test ARG] [--include-external]

  Cabal operations

Available options:
  --discover-all           Discover all import dependencies
  --discover-all-libs      Discover all library import dependencies
  --discover-all-exes      Discover all executable import dependencies
  --discover-all-tests     Discover all test import dependencies
  --no-discover-all-libs   Discover no library import dependencies
  --no-discover-all-exes   Discover no executable import dependencies
  --no-discover-all-exes   Discover no test import dependencies
  --discover-lib           Discover the default library import dependencies
  --no-discover-lib        Don't discover the default library import
                           dependencies
  --discover-sublib ARG    Discover specified sublibrary import dependencies
  --no-discover-sublib ARG Don't discover specified sublibrary import
                           dependencies
  --discover-exe ARG       Discover specified executable import dependencies
  --no-discover-exe ARG    Don't discover specified executable import
                           dependencies
  --discover-test ARG      Discover specified test import dependencies
  --no-discover-test ARG   Don't discover specified test import dependencies
  --include-external       Include external import dependencies
  -h,--help                Show this help text
```

If you pass no `discover` args, it only discovers the main library.

If you do pass `discover` args, they are applied left-to-right. This makes for a nice CLI. For example:

```
# Discover all modules _except_ the graphex exe's
$ graphex cabal --discover-all --no-discover-exe graphex

# Discover everything except tests
$ graphex cabal --discover-all --no-discover-tests
```

I tested the new CLI on graphex and it seems to work. And the test suite for the logic is pretty solid.

### CODE REVIEW

**Comment by dustin** [comment]
This looks pretty neat.  You can push it if you like it.  :)

**Comment by ramirez7** [comment]
I will once I get a little more coverage on the CLI options!

**Comment by ramirez7** [comment]
Oh and I have to actually write the CLI lol.

File: app/Main.hs [hunk 1]
```diff
@@ -23,10 +22,9 @@ import           Options.Applicative  (Parser, argument, command,
                                        switch, value, (<**>))
 
 import           Graphex
-import           Graphex.Cabal
 import           Graphex.Core
 import qualified Graphex.CSV
-import           Graphex.LookingGlass
+import           Main.Cabal
```
Comment (ramirez7) [inline app/Main.hs:27]: I moved the `graphex cabal` CLI to its own module.

File: app/Main/Cabal.hs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  , optIncludeExternal :: Bool
+  } deriving stock Show
+
+justWhen :: a -> Bool -> Maybe a
+justWhen a True  = Just a
+justWhen _ False = Nothing
+
+cabalOptions :: Parser CabalOptions
+cabalOptions = do
+  optToDiscover <- fmap mconcat $ many $ asum
```
Comment (ramirez7) [inline app/Main/Cabal.hs:25]: This collects the options and preserves order. The actual runner then folds them left-to-right and this type's Semigroup instance does the fun stuff.

File: dummy-sublib/DummySublibModule.hs [hunk 3]
```diff
@@ -0,0 +1 @@
+module DummySublibModule where
```
Comment (ramirez7) [inline dummy-sublib/DummySublibModule.hs:1]: I made this module + internal library just for the test suite.

File: src/Graphex/Logger.hs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+module Graphex.Logger (logit) where
```
Comment (ramirez7) [inline src/Graphex/Logger.hs:1]: I didn't commit anything in this module, but I did use it for testing.

It's just a wrapper around `Debug.Trace` with a verbosity controlled by an env var.

File: test/CabalSpec.hs [hunk 5]
```diff
@@ -39,31 +40,58 @@ exeModules :: [Module]
 exeModules =
   [ Module "Paths_graphex" ModuleNoFile
   , Module "graphex-Main" "app/Main.hs"
+  , Module "Main.Cabal" "app/Main/Cabal.hs"
```
Comment (ramirez7) [inline test/CabalSpec.hs:43]: Using graphex itself as a test is fun.

**Pull request merged**